pkg/controller/clusterpool/collections.go

```diff
@@ -584,6 +617,475 @@ func (cds *cdCollection) Delete(c client.Client, cdName string) error {
+	sort.Slice(
+		cdcs.unassigned,
+		func(i, j int) bool {
+			now := metav1.NewTime(time.Now())
+			iStatus := conditionsv1.FindStatusCondition(cdcs.unassigned[i].Status.Conditions, hivev1.ApplySucceededCondition)
+			jStatus := conditionsv1.FindStatusCondition(cdcs.unassigned[j].Status.Conditions, hivev1.ApplySucceededCondition)
+			iName := cdcs.unassigned[i].Name
+			jName := cdcs.unassigned[j].Name
+			if iStatus == nil || iStatus.Status == corev1.ConditionUnknown {
+				iStatus = &conditionsv1.Condition{Reason: hivev1.CustomizationApplyReasonSucceeded}
+				iStatus.LastTransitionTime = now
+			}
+			if jStatus == nil || jStatus.Status == corev1.ConditionUnknown {
+				jStatus = &conditionsv1.Condition{Reason: hivev1.CustomizationApplyReasonSucceeded}
+				jStatus.LastTransitionTime = now
+			}
+			iTime := iStatus.LastTransitionTime
+			jTime := jStatus.LastTransitionTime
+			if iStatus.Reason == jStatus.Reason {
+				if iTime.Equal(&jTime) {
+					// Sort by name to make this deterministic
+					return iName < jName
+				}
+				return iTime.Before(&jTime)
+			}
+			if iStatus.Reason == hivev1.CustomizationApplyReasonSucceeded {
+				return true
+			}
+			if jStatus.Reason == hivev1.CustomizationApplyReasonSucceeded {
+				return false
+			}
+			return iName < jName
+		},
+	)
+}
+
+func (cdcs *cdcCollection) Reserve(c client.Client, cdc *hivev1.ClusterDeploymentCustomization, cdName, poolName string) error {
+	cdc.Status.ClusterDeploymentRef = &corev1.LocalObjectReference{Name: cdName}
+	cdc.Status.ClusterPoolRef = &corev1.LocalObjectReference{Name: poolName}
+
+	changed := conditionsv1.SetStatusConditionNoHeartbeat(&cdc.Status.Conditions, conditionsv1.Condition{
+		Type:    conditionsv1.ConditionAvailable,
+		Status:  corev1.ConditionFalse,
+		Reason:  "Reserved",
+		Message: "reserved",
+	})
+
+	if changed {
+		if err := c.Status().Update(context.Background(), cdc); err != nil {
+			return err
+		}
+	}
+
+	cdcs.reserved[cdc.Name] = cdc
+	cdcs.byCDCName[cdc.Name] = cdc
+
+	for i, cdci := range cdcs.unassigned {
+		if cdci.Name == cdc.Name {
+			copy(cdcs.unassigned[i:], cdcs.unassigned[i+1:])
+			cdcs.unassigned = cdcs.unassigned[:len(cdcs.unassigned)-1]
+			break
+		}
+	}
+
+	cdcs.Sort()
+	return nil
+}
+
+func (cdcs *cdcCollection) Unassign(c client.Client, cdc *hivev1.ClusterDeploymentCustomization) error {
+	changed := conditionsv1.SetStatusConditionNoHeartbeat(&cdc.Status.Conditions, conditionsv1.Condition{
+		Type:    conditionsv1.ConditionAvailable,
+		Status:  corev1.ConditionTrue,
+		Reason:  "Available",
+		Message: "available",
+	})
+
+	if cdc.Status.ClusterDeploymentRef != nil || cdc.Status.ClusterPoolRef != nil {
+		cdc.Status.ClusterDeploymentRef = nil
+		cdc.Status.ClusterPoolRef = nil
+		changed = true
+	}
+
+	if changed {
+		if err := c.Status().Update(context.Background(), cdc); err != nil {
+			return err
+		}
+	}
+
+	if _, ok := cdcs.reserved[cdc.Name]; ok {
+		delete(cdcs.reserved, cdc.Name)
+	}
```

Comment:
`delete` is idempotent.
```suggestion
	delete(cdcs.reserved, cdc.Name)
```